yarn-project/ethereum/src/l1_tx_utils.ts

```diff
@@ -0,0 +1,347 @@
+                throw err;
+              }
+            }
+          }
+        }
+
+        const tx = await this.publicClient.getTransaction({ hash: currentTxHash });
+        const timePassed = Date.now() - lastSeen;
+
+        if (tx && timePassed < gasConfig.stallTimeMs!) {
+          this.logger?.debug(`L1 Transaction ${currentTxHash} pending. Time passed: ${timePassed}ms`);
+          await sleep(gasConfig.checkIntervalMs!);
+          continue;
+        }
+
+        if (timePassed > gasConfig.stallTimeMs! && attempts < gasConfig.maxAttempts!) {
+          attempts++;
+          const newGasPrice = await this.getGasPrice(gasConfig, attempts);
+
+          this.logger?.debug(
+            `L1 Transaction ${currentTxHash} appears stuck. Attempting speed-up ${attempts}/${gasConfig.maxAttempts} ` +
+              `with new priority fee ${formatGwei(newGasPrice.maxPriorityFeePerGas)} gwei`,
+          );
+
+          currentTxHash = await this.walletClient.sendTransaction({
+            ...request,
+            nonce,
+            gas: params.gasLimit,
+            maxFeePerGas: newGasPrice.maxFeePerGas,
+            maxPriorityFeePerGas: newGasPrice.maxPriorityFeePerGas,
+          });
+
+          txHashes.add(currentTxHash);
+          lastSeen = Date.now();
+        }
+        await sleep(gasConfig.checkIntervalMs!);
+      } catch (err: any) {
+        this.logger?.warn(`Error monitoring tx ${currentTxHash}:`, err);
+        if (err.message?.includes('reverted')) {
+          throw err;
+        }
+        await sleep(gasConfig.checkIntervalMs!);
+      }
+      // Check if tx has timed out.
+      if (gasConfig.txTimeoutMs) {
+        txTimedOut = Date.now() - initialTxTime > gasConfig.txTimeoutMs!;
+      }
+    }
+    throw new Error(`L1 Transaction ${currentTxHash} timed out`);
+  }
+
+  /**
+   * Sends a transaction and monitors it until completion
+   * @param request - The transaction request (to, data, value)
+   * @param gasConfig - Optional gas configuration
+   * @returns The receipt of the successful transaction
+   */
+  public async sendAndMonitorTransaction(
+    request: L1TxRequest,
+    gasConfig?: Partial<L1TxUtilsConfig>,
+  ): Promise<TransactionReceipt> {
+    const { txHash, gasLimit } = await this.sendTransaction(request, gasConfig);
+    return this.monitorTransaction(request, txHash, { gasLimit }, gasConfig);
+  }
+
+  /**
+   * Gets the current gas price with bounds checking
+   */
+  private async getGasPrice(_gasConfig?: L1TxUtilsConfig, attempt: number = 0): Promise<GasPrice> {
+    const gasConfig = { ...this.config, ..._gasConfig };
+    const block = await this.publicClient.getBlock({ blockTag: 'latest' });
+    const baseFee = block.baseFeePerGas ?? 0n;
+
+    // Get initial priority fee from the network
+    let priorityFee = await this.publicClient.estimateMaxPriorityFeePerGas();
+    let maxFeePerGas = gasConfig.maxGwei! * WEI_CONST;
```

Comment:
This means that the user specified `maxGwei` will not be the maximum, but the value used for the first attempt.

I'd suggest calculating the maxFeePerGas as baseFee (fetched from the `getBlock` with some padding, keeping in mind that baseFee is capped to increase no more than 12.5% per block, so based on how fast you plan to replace you know up to how much base fee can climb until then) plus whatever you choose as priority fee. Then, after all bumps, check if you've hit max, and if so, bail.

Reply:
right, first part makes sense as we end up bumping the specified `maxGwei`
confused about the 2nd part with the 12.5% increase. We run this on each attempt so it will fetch the new block's baseFee, why do we need to calculate this 12.5% max increase per block?

Reply:
Sorry, re-reading my comment I see I wasn't very clear. What I meant is the following: let's say base fees are increasing consistently block over block, and the priority fee is zero (for simplicity). So if block N is at a base fee of 100, then:
```
N: 100
N+1: 112
N+2: 127
N+3: 142
```

Now, let's say that `stallTime` is 24s (2 L1 blocks). If you sent the tx at block N with a base fee of exactly 100, then your tx will not be elligible for inclusion at N+1, and you'll have to wait until N+2 to read the new base fee and resubmit.

If instead you submit your tx at N with a base fee of 127, you know that your tx will always be elligible until the time you retry, at which point you just bump the priority fee (and the total fee to make geth happy). This way you guarantee that you don't "miss" the opportunity of going in N+1.